main/platform/espWifi.cpp

```diff
@@ -0,0 +1,396 @@
+            return false;
+        }
+
+        const auto ssid = mainNs->getString(KeyWifiApSsid, "Jaculus-esp32");
+        const auto pass = mainNs->getString(KeyWifiApPass);
+
+        wifi_config_t cfg = {};
+        if (pass.size() >= 8) {
+            snprintf((char*)cfg.ap.password, 64, "%s", pass.c_str());
+            cfg.ap.authmode = WIFI_AUTH_WPA2_PSK;
+        } else {
+            if(pass.size() != 0) {
+                jac::Logger::error("The WiFi password is too short, 8 characters required, leaving the WiFI open!");
+            }
+            cfg.ap.authmode = WIFI_AUTH_OPEN;
+        }
+        snprintf((char*)cfg.ap.ssid, 32, "%s", ssid.c_str());
+        cfg.ap.beacon_interval = 400;
+        cfg.ap.max_connection = 4;
+
+        err = esp_wifi_set_config(WIFI_IF_AP, &cfg);
+        if(err != ESP_OK) {
+            jac::Logger::error("esp_wifi_set_config: " + std::string(esp_err_to_name(err)));
+            return false;
+        }
+        break;
+    }
+    default:
+        break;
+    }
+
+    auto err = esp_wifi_start();
+    if(err != ESP_OK) {
+        jac::Logger::error("esp_wifi_start: " + std::string(esp_err_to_name(err)));
+        return false;
+    }
+
+    return true;
+}
+
+void EspWifiController::eventHandlerWifi(void* selfVoid, esp_event_base_t event_base,
+        int32_t event_id, void* event_data) {
+    auto *self = (EspWifiController*)selfVoid;
+
+    switch(event_id) {
+        case WIFI_EVENT_STA_START:
+        case WIFI_EVENT_STA_DISCONNECTED: {
+            std::lock_guard<std::mutex> lg(self->_modeLock);
+
+            if(self->_mode != Mode::STATION) {
+                break;
+            }
+
+            if(self->_staMode == StaMode::SPECIFIC_SSID) {
+                auto err = esp_wifi_connect();
+                if(err != ESP_OK) {
+                    jac::Logger::error("esp_wifi_connect: " + std::string(esp_err_to_name(err)));
+                }
+            } else if(self->_staMode == StaMode::BEST_SIGNAL) {
+                auto err = esp_wifi_scan_start(NULL, false);
+                if(err != ESP_OK) {
+                    jac::Logger::error("esp_wifi_scan_start: " + std::string(esp_err_to_name(err)));
+                }
+            }
+            break;
+        }
+        case WIFI_EVENT_SCAN_DONE: {
+            std::lock_guard<std::mutex> lg(self->_modeLock);
+
+            if(self->_mode != Mode::STATION || self->_staMode != StaMode::BEST_SIGNAL) {
+                esp_wifi_clear_ap_list();
+                break;
+            }
+
+            auto wifiNs = EspNvsKeyValue::open(KvNsWifiSsid);
+            if(!wifiNs) {
+                esp_wifi_clear_ap_list();
+                break;
+            }
+
+            wifi_ap_record_t rec;
+            const std::string not_exists_sentinel = "a";
+            while(true) {
+                auto err = esp_wifi_scan_get_ap_record(&rec);
```

Comment:
`esp_wifi_scan_get_ap_record` is not available in esp-idf 5.1

Reply:
It is in 5.1.4

Reply:
it seemed like a quick change but updating idf seems more reasonable